_posts/2016-03-16-AngularTS.md

```diff
@@ -0,0 +1,413 @@
+It is possible that you have to declare the referencing on the top of your file.
+{% highlight bash %}
+/// <reference path="../../typings/angularjs/angular.d.ts" />
+/// <reference path="../../typings/angularjs/angular-route.d.ts" />
+{% endhighlight %}
+
+## Modules
+
+### Angular Modules
+Modules are here to help us modularize our code.
+But it is important to know that if you’re not planning to share or declare related functionality.
+It is best practice to use only one main module as the root of your application. 
+To let the module know the existence of every component, they have to register themselves.
+Below every component declaration you will see a registration to the module. 
+When registering the module you have to add all the libraries you want to depend upon.
+In this example we inject the routing service for navigation. 
+
+{% highlight javascript %}
+module JWorks {
+    "use strict";
+
+    angular
+       .module("jworks360", ['ngRoute'])
+     }
+{% endhighlight %}
+
+### Internal TypeScript Modules
+These modules are similar to namespaces.
+You can define an unique namespace around your code.
+This will encapsulate variables, interfaces, and classes. 
+TypeScript supports sub namespaces for further encapsulation.
+
+{% highlight javascript %}
+module JWorks {
+    "use strict";
+}
+{% endhighlight %}
+
+#### Transpiled JavaScript
+
+{% highlight javascript %}
+var JWorks;
+(function(JWorks){
+
+})(JWorks || (JWorks = {}));
+{% endhighlight %}
+
+To encapsulate our code, the module will transpile to an IIFE around our components.
+This will avoid global code which helps prevent variables and function declarations from living longer than expected in the global scope, which also helps avoid variable collisions.
+
+
+## Entity Class
+Now that TypeScript supports OO, we can analyse our business problem and define the business objects into entity classes.
+When you analyse and define these entities you can define which properties and methods each entity needs.
+If you have a couple entities, you can even establish a relationship.
+This will provide a clear view on what you want to achieve and have the possibility to create multiple instances of these classes. 
+When building an entity class you can optionally define an interface to show what’s the intent. 
+
+{% highlight javascript %}
+module JWorks {
+
+    export interface IEmployee {
+         username:string;
+         name:string;
+         eat(food:string):void
+    }
+
+    export class Employee implements IEmployee {
+ 
+          constructor(public username:string, public name:string){}
+
+          eat(food:string):void{
+             //implementation
+          }
+    }
+}
+
+{% endhighlight %}
+
+To use your entity class in a controller you have to define the export key.
+This will expose the class to other classes.
+When exporting the interface you will use it as a data type.
+
+### Class as property
+{% highlight javascript %}
+Employee :IEmployee;
```

Comment:
Remove the space before the colon in order to be consistent with the other code samples?